cpp/src/spatial/polygon_distance.cu

```diff
@@ -0,0 +1,161 @@
+    // if point is projected outside line segment bounds, use point-point distance.
+    if (0 < travel && travel < magnitude) {  // 0 < travel < edge_length
+      return abs(point_y * edge_x - point_x * edge_y) * rhypot(edge_x, edge_y);
+    } else {
+      return hypot(point_x, point_y);
+    }
+  }
+};
+
+struct directed_polygon_distance_functor {
+  template <typename T, typename... Args>
+  std::enable_if_t<not std::is_floating_point<T>::value, std::unique_ptr<cudf::column>> operator()(
+    Args&&...)
+  {
+    CUSPATIAL_FAIL("Non-floating point operation is not supported");
+  }
+
+  template <typename T>
+  std::enable_if_t<std::is_floating_point<T>::value, std::unique_ptr<cudf::column>> operator()(
+    cudf::column_view const& xs,
+    cudf::column_view const& ys,
+    cudf::column_view const& space_offsets,
+    rmm::mr::device_memory_resource* mr,
+    cudaStream_t stream)
+  {
+    size_type num_points  = xs.size();
+    size_type num_spaces  = space_offsets.size();
+    size_type num_results = num_spaces * num_spaces;
+
+    if (num_results == 0) {
+      return cudf::make_empty_column(cudf::data_type{cudf::type_to_id<T>()});
+    }
+
+    // ===== Make Separation and Key Iterators =====================================================
+
+    auto cartesian_iter = make_cartesian_product_group_index_iterator(
+      num_points, num_spaces, space_offsets.begin<cudf::size_type>());
+
+    auto cartesian_key_iter = thrust::make_transform_iterator(
+      cartesian_iter, [] __device__(cartesian_product_group_index idx) {
+        return thrust::make_pair(idx.group_a.idx, idx.group_b.idx);
+      });
+
+    auto d_xs = cudf::column_device_view::create(xs);
+    auto d_ys = cudf::column_device_view::create(ys);
+
+    auto separation_iter = thrust::make_transform_iterator(
+      cartesian_iter, segment_point_distance_calculator<T>{*d_xs, *d_ys});
+
+    // ===== Materialize ===========================================================================
+
+    auto result = cudf::make_fixed_width_column(cudf::data_type{cudf::type_to_id<T>()},
+                                                num_results,
+                                                cudf::mask_state::UNALLOCATED,
+                                                stream,
+                                                mr);
+
+    auto num_cartesian = num_points * num_points;
+
+    thrust::reduce_by_key(rmm::exec_policy(stream)->on(stream),
+                          cartesian_key_iter,
+                          cartesian_key_iter + num_cartesian,
+                          separation_iter,
+                          thrust::make_discard_iterator(),
+                          result->mutable_view().begin<T>(),
+                          thrust::equal_to<thrust::pair<int32_t, int32_t>>(),
+                          thrust::minimum<T>());
+
+    return result;
+  }
+};
+
+}  // namespace
+}  // namespace detail
+
```

Comment:
Typically we have a detail version of API functions that take a stream, like we do in libcudf, don't we?

Reply:
Probably. I'll fix this PR and then go take ensure it's being done elsewhere in a followup (if needed).